File: src/main/java/org/funivan/lologa/tiles/Tiles.java
package org.funivan.lologa.tiles;

import org.funivan.lologa.tile.Position.PositionInterface;
import org.funivan.lologa.tile.TileInterface;

import java.util.HashMap;

public class Tiles implements TilesInterface {
    final HashMap<PositionInterface, TileInterface> tiles;

    public Tiles(HashMap<PositionInterface, TileInterface> tiles) {
        this.tiles = tiles;
    }

    public Tiles() {
        this(new HashMap<>());
    }


    @Override
    public TileInterface get(PositionInterface position) {
        return this.tiles.get(position);
    }

    @Override
    public boolean has(PositionInterface position) {
        return this.tiles.containsKey(position);
    }

    @Override
    public TilesInterface without(PositionInterface position) {
        HashMap<PositionInterface, TileInterface> result = this.tiles;
        result.remove(position);
        return new Tiles(result);
    }

    @Override
    public TilesInterface with(TileInterface tile) {
        HashMap<PositionInterface, TileInterface> result = this.tiles;
        result.put(tile.position(), tile);
        return new Tiles(result);
    }

    @Override
    public Iterable<TileInterface> all() {
        return this.tiles.values();
    }

}

File: src/main/java/org/funivan/lologa/tiles/Tiles.kt
package org.funivan.lologa.tiles

import org.funivan.lologa.tile.position.PositionInterface
import org.funivan.lologa.tile.TileInterface
import java.util.*

class Tiles(private val tiles: HashMap<PositionInterface, TileInterface> = HashMap()) : TilesInterface {

    override fun get(position: PositionInterface): TileInterface {
        var tile = this.tiles.get(position);
        return tile ?: throw Exception("Invalid tile position " + position.toString())
    }

    override fun has(position: PositionInterface): Boolean {
        return this.tiles.containsKey(position)
    }

    override fun without(position: PositionInterface): TilesInterface {
        val result = HashMap(this.tiles)
        result.remove(position)
        return Tiles(result)
    }

    override fun with(tile: TileInterface): TilesInterface {
        val result = HashMap(this.tiles)
        result.put(tile.position(), tile)
        return Tiles(result)
    }

    override fun all(): Iterable<TileInterface> {
        return this.tiles.values
    }

    override fun size(): Int {
        return this.tiles.size
    }
}
